confgenerator/fluentbit/parser_deduplication.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fluentbit
+
+const (
+	ParserNestLuaFunction = `parser_nest`
+
+	// ParserNestLuaScriptContents is an incomplete Lua funtion that is completed when populated
+	// with the parse_key (i.e. the key being parsed).
+	ParserNestLuaScriptContents = `
+  function parser_nest(tag, timestamp, record)
+  record["logging.googleapis.com/__tmp"] = {}
+  
+  for k, v in pairs(record) do
+      if k ~= "%s" then
+          if record["logging.googleapis.com/__tmp"] == nil then
+              record["logging.googleapis.com/__tmp"] = {}
+          end
+          record["logging.googleapis.com/__tmp"][k] = v
+          record[k] = nil
+      end
+  end
+
+  return 2, timestamp, record
+end
+`
+	ParserMergeLuaFunction       = `parser_merge_record`
+	ParserMergeLuaScriptContents = `
+  function shallow_merge(record, parsedRecord)
+    -- If no exiting record exists
+    if (record == nil) then 
+        return parsedRecord
+    end
+    
+    for k, v in pairs(parsedRecord) do
+        record[k] = v
+    end
+
+    return record
+end
+
+function merge(record, parsedRecord)
+    -- If no exiting record exists
+    if record == nil then 
+        return parsedRecord
+    end
+    
+    -- Potentially overwrite or merge the original records.
+    for k, v in pairs(parsedRecord) do
+        -- If there is no conflict
+        if k == "logging.googleapis.com/logName" then 
+            -- Ignore the parsed payload
```

Comment:
There were a few reasons. We have an internal ticket tracking the effort to make lognames more controllable by the Ops Agent (right now it can only be controlled by the receiver_id) [b/219991181](https://b.corp.google.com/issues/219991181), but it's a bit tricky since:

- The `out_stackdriver` exporter in Fluent Bit adds a prefix to the logname (`projects/<project_id>/logs/`) and since that is the user visible logname, it requires some more thought into how we should support that use case. We can move somee of this discussion over https://github.com/GoogleCloudPlatform/ops-agent/pull/419 since technically after this change, we can make progress over there.
- Currently doesn't matter what is parsed, the ops Agent overwrites the logname at the end now. Made sense to make explicit the behaviour here and in documentation until https://github.com/GoogleCloudPlatform/ops-agent/pull/419 is resolved.

Reply:
Makes sense